invenio_notifications/services/builders.py

```diff
@@ -0,0 +1,77 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 CERN.
+# Copyright (C) 2023 Graz University of Technology.
+#
+# Invenio-Records-Resources is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""Errors."""
+
+from abc import abstractmethod
+from typing import Dict
+
+from invenio_notifications.models import Notification, Recipient
+from invenio_notifications.services.filters import RecipientFilter
+from invenio_notifications.services.generators import (
+    ContextGenerator,
+    RecipientBackendGenerator,
+    RecipientGenerator,
+)
+
+
+class NotificationBuilder:
+    """Base notification builder."""
+
+    context: list[ContextGenerator]
+    recipients: list[RecipientGenerator]
+    recipient_filters: list[RecipientFilter]
+    recipient_backends: list[RecipientBackendGenerator]
+
+    type = "Notification"
+
+    @classmethod
+    @abstractmethod
+    def build(cls, **kwargs):
+        """Build notification based on type and additional context."""
+
+        return Notification(
+            type=type,
+            context={},
+        )
```

Comment:
```suggestion
    @abstractmethod
    def build(cls, **kwargs):
        """Build notification based on type and additional context."""
        raise NotImplemented()
```

minor: since it's an `abstractmethod`. Ditto for all similar cases, e.g. in `generators.py`